code/game/objects/structures/morgue.dm

```diff
@@ -242,67 +247,69 @@ GLOBAL_LIST_EMPTY(crematoriums)
 		else
 			src.icon_state = "crema1"
 
-		if(locked)
+		if(cremating)
 			src.icon_state = "crema_active"
 
 	return
 
-/obj/structure/bodycontainer/crematorium/proc/cremate(mob/user)
-	if(locked)
-		return //don't let you cremate something twice or w/e
-	// Make sure we don't delete the actual morgue and its tray
-	var/list/conts = GetAllContents() - src - connected
+/obj/structure/bodycontainer/crematorium/process()
+	if(cremating)
+		cremate()
 
-	if(!conts.len)
-		audible_message("<span class='hear'>You hear a hollow crackle.</span>")
-		return
-
-	else
-		audible_message("<span class='hear'>You hear a roar as the crematorium activates.</span>")
+///Toggles cremation. The cremate parameter can be used to set a specific value instead of toggling.
+/obj/structure/bodycontainer/crematorium/proc/toggle_cremation(set_cremate, mob/user)
+	if(cremating && set_cremate != TRUE)
+		cremating = FALSE
+		locked = FALSE
+		STOP_PROCESSING(SSobj, src)
+		update_icon()
 
+		audible_message("<span class='hear'>You hear a roar as [src] activates.</span>")
+		return
+	if(!cremating && set_cremate != FALSE)
+		if(user)
+			last_user = user
+		else
+			last_user = null //let's not blame the previous guy
+		cremating = TRUE
+		first_cremation_tick = TRUE
 		locked = TRUE
+		START_PROCESSING(SSobj, src)
 		update_icon()

```

Comment:
are the cases backwards for the audible messages? looks like you're doing the turn off one when it turns on and vice versa

also why do ``set_cremate != FALSE`` and same for true, why not just ``set_cremate`` and ``!set_cremate``?

also also if !user is going to be null, why not just last_user = user

Reply:
1) they're backwards, will fix
2) my intention was to essentially make it skip the activation/deactivation is set_cremate would prevent it, while also ignoring the value if it's null. I'm not sure right now if `!= FALSE` would also catch null though, that would break the check.
3) As the comment implies, it makes sure that if user A cremates someone and then the crematorium is activated again later without an user it doesn't get logged as A doing it

Reply:
on 2 I can see where you're coming from I suppose

on 3 if its activated without a user won't user be null and hence you can just set last_user to user (which should be null in that case?)

Reply:
oh, true, didn't consider that